src/adapter/src/catalog/storage.rs

```diff
@@ -566,6 +577,86 @@ async fn migrate(
 
             Ok(())
         },
+        // Namespacing database ids and schema ids by User or System. Currently
+        // everything exists in the "User" schema.
+        //
+        // Introduced in v0.51.0
+        //
+        // TODO(parkertimmerman): We can make DatabaseKey and SchemaKey more idomatic
+        // enums after this upgrade.
+        |txn: &mut Transaction<'_>, _now, _bootstrap_args| {
+            // Migrate all of our DatabaseKeys.
+            txn.databases.migrate(|key, value| {
```

Comment:
The new struct fields and this migration can maybe be done in an easier way, if done *before* the big vec of migrations is run. You could write a stash function that calls iter_raw on these and attempts to marshal into the old struct. If successful, then you add a retraction to a batch and an addition with the new key. If the marshal into the old struct fails, it's already the new struct and should be skipped. The old structs would be copied and exist only within this migration function, and everything could then use the new enums elsewhere. Does that work? I think it'd simplify this.

Reply:
Being able to have `enum DatabaseKey { User(u64), System(u64) }`, and a `enum SchemaKey` of the same, would definitely simplify things! My hesitation to adding custom migration logic like this, or custom serde impls was to prevent any complications when we later try to improve migrations, i.e. https://github.com/MaterializeInc/materialize/issues/17466. Although off the top of my head I don't think what you're proposing should complicate anything too much, so I don't really feel strongly one way or another, so I'll defer to you here on what you think is best :)